### PR TITLE
[WIP] Make `Command` class "stateless"

### PR DESCRIPTION
**Pull request by veox**
WIP, demo.

Closes #175 (is part of this PR).
Closes #193 (different solution).

See linked PRs for earlier discussion, and some of my arguments.

### What was wrong?

* Every peer's got a `Protocol` (or sub-class) instance, and every instance of that holds (identical) instances of `Command`s in the protocol.
* `cmd_offset` and `snappy_support` - protocol implementation detail and per-connection protocol feature, respectively - are passed around needlessly, requiring every sub-protocol developer to know of their existence.

### How was it fixed?

Only dirty so far.

* A module-level "look-up table" for `Command`s and a function to operate it are introduced.
* `Command`s are referenced by class.
* Each `Protocol` instance maintains a list of command-plus-feature classes they support, plus `dict` mappings (similarly to how it was done previously).
* `Protocol`s are still per-peer objects. This can be addressed later.

TODO:

- [ ] `BCC` and `LES` need to be updated, too!
- [ ] Run tests locally and fix failures.
- [ ] Address TODOs and FIXMEs in code.

### Cute Animal Picture

put a cute animal picture link inside the parentheses

Source: [Pumpkin the raccoon](https://cuteandkids.com/inspire/pumpkin-the-raccoon/), by Laura Young

### CODE REVIEW

File: p2p/peer.py [hunk 1]
```diff
@@ -301,12 +299,12 @@ def remove_subscriber(self, subscriber: 'PeerSubscriber') -> None:
         """
         await self.send_sub_proto_handshake()
         cmd, msg = await self.read_msg()
-        if isinstance(cmd, Ping):
+        if cmd.cmd_type == Ping:
```
Comment (veox) [inline p2p/peer.py]: This (here and throughout) is the first very-WIP'y thing. `cmd` is now a class, not an instantiated object of that class.

File: p2p/protocol.py [hunk 2]
```diff
@@ -159,6 +164,37 @@ def encode(self, data: PayloadType) -> Tuple[bytes, bytes]:
         body = _pad_to_16_byte_boundary(enc_cmd_id + compressed_payload)
         return header, body
 
+### FIXME: move to p2p._utils module?..
```
Comment (veox) [inline p2p/protocol.py]: Here and below, module-level stuff that's probably undesirable.

File: p2p/protocol.py [hunk 3]
```diff
@@ -159,6 +164,37 @@ def encode(self, data: PayloadType) -> Tuple[bytes, bytes]:
...
+# FIXME: use kwargs for feature specification; they shouldn't be spelled out here!
+def get_command_class(cmd_class, cmd_id_offset, snappy_support):
+    # TODO: use NamedTuple?..
+    specifier = (cmd_class, cmd_id_offset, snappy_support)
+
+    # use existing if available
+    if specifier in command_classes.keys():
+        return command_classes[specifier]
+
+    class CommandClassInstance(cmd_class):
```
[the remark this answers is not in the document]
Comment (veox) [reply p2p/protocol.py]: Comment outdated because name of class changed.

I did try to use `metaclass` (it didn't work, not with the little effort I gave); but now think that doing `type(cmd)` elsewhere in the code would obfuscate that:

* `cmd` now refers to a class, not an object; in fact, it's a sub-class of what would've previously been returned by `type(cmd)`;
* trying to use return value of `type(cmd)` directly to create another variable referring to same class `cmd` would not yield the expected result.

In short, this would make things even more confusing than already. I'm not well-versed to pursue this.

File: p2p/peer.py [hunk 4]
```diff
@@ -488,23 +488,20 @@ def process_msg(self, cmd: protocol.Command, msg: protocol.PayloadType) -> None:
...
+        # Snappy compression is negotiated indirectly, by comparing protocol versions;
+        # if both sides support v5, then compression is mandatory
         snappy_support = msg['version'] >= SNAPPY_PROTOCOL_VERSION
-
         if snappy_support:
-            # Now update the base protocol to support snappy compression
-            # This is needed so that Trinity is compatible with parity since
-            # parity sends Ping immediately after Handshake
-            self.base_protocol = P2PProtocol(self, snappy_support=snappy_support)
+            # TODO: select protocol directly, instead of using side-effect of property
```
Comment (pipermerriam) [inline p2p/peer.py]: :+1:

File: p2p/protocol.py [hunk 5]
```diff
@@ -55,94 +55,99 @@ class TypedDictPayload(TypedDict):
...
-        self.cmd_id = cmd_id_offset + self._cmd_id
-        self.snappy_support = snappy_support
-
+    # FIXME: does this even work?
     @property
-    def logger(self) -> logging.Logger:
-        if self._logger is None:
-            self._logger = logging.getLogger(f"p2p.protocol.{type(self).__name__}")
-        return self._logger
+    def logger(cls) -> logging.Logger:
```
Comment (pipermerriam) [inline p2p/protocol.py]: This won't work if we're interacting with these as classes.  It'll need to be converted to a static property **or** a metaclass if we want to get fancy and have the logger instance be a class variable but not have to manually set it each time.

File: p2p/protocol.py [hunk 6]
```diff
@@ -55,94 +55,99 @@ class TypedDictPayload(TypedDict):
...
-    @property
-    def is_base_protocol(self) -> bool:
-        return self.cmd_id_offset == 0
+    # TODO: add abstract methods for is_base_protocol
 
-    def __str__(self) -> str:
-        return f"{type(self).__name__} (cmd_id={self.cmd_id})"
+    # FIXME: reference to `self`
+    # def __str__(self) -> str:
+    #     return f"{type(self).__name__} (cmd_id={cls.cmd_id})"
```
Comment (pipermerriam) [inline p2p/protocol.py]: This might require a metaclass which I don't see being a problem.

https://stackoverflow.com/questions/8144026/how-to-define-a-str-method-for-a-class

File: p2p/protocol.py [hunk 7]
```diff
@@ -159,6 +164,38 @@ def encode(self, data: PayloadType) -> Tuple[bytes, bytes]:
...
+# FIXME: use kwargs for feature specification; they shouldn't be spelled out here!
+def get_command_class(cmd_class, cmd_id_offset, snappy_support):
+    # TODO: use NamedTuple?..
+    specifier = (cmd_class, cmd_id_offset, snappy_support)
+
+    # use existing if available
+    if specifier in command_classes.keys():
+        return command_classes[specifier]
+
+    class CommandWithSettings(cmd_class):
```
Comment (pipermerriam) [inline p2p/protocol.py]: Dynamically generated classes are going to make type checking incredibly hard and this strikes me as equivalent to using instances of the command class so it seems like it defeats the purpose of this pull request since now you're effectively replacing instances with dynamically constructed classes...

Comment (veox) [reply p2p/protocol.py]: This is somewhat tangential, but objects of this class don't have to be re-instantiated on every new connection (via `Protocol` (via `Peer`)) - so they no longer are. It is the same in-memory object (describing a class and its class methods), referenced through thousands of connections within a lifetime of a node run (mostly ending prematurely with "wrong network" and "too many peers"). So, at the very least, it's not equivalent.

"Tangential", because while striving for `snappy_support` to not be passed around in [sub-]protocol implementations, this PR passes around references to some `Command` class in general. The implementing `Protocol` then looks up a sub-class of the `Command` that it's supposed to be _always_ using (with appropriate `cmd_id_offset` and `snappy_support`).

The purpose of this PR is to stop referencing [sub-]protocol features when issuing commands belonging to that [sub-]protocol; and to stop creating thousands of identical `Command` objects; in fact, stop creating `Command` objects altogether.

File: p2p/peer.py (hunk 4 above)
Comment (veox) [reply p2p/peer.py]: I don't think this can be done in a meaningful manner until `Protocol`s have their own `get_protocol_class()` similar to this PR's `get_command_class()`.

To do it right now, the existing line can be kept. The difference is only that previously, a new `P2PProtocol` object must be created, whereas with this PR, only most of it (the property and the command look-up tables). That, and the fact that "P2PProtocol" features again (compare to "whatever is already in use"), and the `self` is passed.

File: p2p/peer.py (hunk 1 above)
Comment (veox) [reply p2p/peer.py]: This line is currently testing for the sub-class. The generic, "protocol-unbound" class, e.g.:

`<class 'trinity.protocol.eth.commands.Transactions'>`

... is [saved as a class attribute](https://github.com/ethereum/trinity/pull/205/files#diff-631c68b238129bf0fb825fe6abb229cbR204) when getting a "protocol-bound" variant of the command with `cmd_id_offset=16` (always the same number for a specific protocol) and `snappy_support=[True|False]` (depending on whether compression is enabled). This "protocol-bound" class is currently ungracefully `__repr__`'d as:

 `<class 'trinity.protocol.eth.commands.Transactions'><cmd_id_offset=16,snappy_support=`[True|False]>`)

Point is, it's not exact matching being tested here, intentionally.

File: p2p/protocol.py (hunk 7 above)
Comment (pipermerriam) [reply p2p/protocol.py]: @veox evidently I never submitted this the other day.  Take a look at https://github.com/ethereum/py-wasm/blob/5b6b61790fc226ff069ecdf179857647113703df/wasm/_utils/singleton.pyfor a working implementation.

Comment (veox) [reply p2p/protocol.py]: Yes, that seems much more concise.

I've decided to lay this off myself for now, as I didn't feel competent enough; but I could try plugging that in (and undo the jolly "why, we'll be passing classes around instead"), if no one else does.

**Comment by veox** [comment]
Probably best to close this, since it's pretty stale; and might be easier to pull off after the plans described in https://github.com/ethereum/trinity/issues/90#issuecomment-488193338 are settled/implemented.
